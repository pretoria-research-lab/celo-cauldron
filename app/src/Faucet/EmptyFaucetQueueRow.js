import React, { useState } from "react";
import PropTypes from "prop-types";
import * as web3Utils from "web3-utils";

const isValid = (address) => {
	return web3Utils.isAddress(address);
};

EmptyFaucetQueueRow.propTypes = () => { 
	return { 
		onClick: PropTypes.func
	}; 
};	

export default function EmptyFaucetQueueRow(props) {

	const [ address, setAddress] = useState("0x0000000000000000000000000000000000000000");
	const { onClick } = props;	

	return (
		<tr>			
			<td><input name="address" onChange={(event) => setAddress(event.target.value)} className="form-control address" value={address} id="address" type="text"></input></td>
			<td colSpan="4"><p>Enter a new faucet request here</p></td>
			<td><button disabled={!isValid(address)} onClick={() => onClick(address)}>{isValid(address) ? "Request" : "Invalid Address"}</button></td>
		</tr>			
	);
}