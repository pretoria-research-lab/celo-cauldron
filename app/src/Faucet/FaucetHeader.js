import React from "react";
import PropTypes from "prop-types";
import airdropIcon from "../assets/13-Airdrop Green Coin (dark bg).png";

FaucetHeader.propTypes = () => { 
	return { 
		network: PropTypes.any
	}; 
};	

export default function FaucetHeader(props) {

	return (
		<>
			<div className="row centered">
				<div className="col-sm-4">
					<div className="row page-header">
						<img className="celoIcons" src={airdropIcon} alt="Airdrop icon" /> 
						<h2 className="mt">{"Celo " + props.network + " Faucet"}</h2>
					</div>
				</div>
				<div className="col-sm-8">
					<div className="row">
						<p>{"A community funded faucet to receive nominal amounts of CELO for testing and development on the Celo " + props.network + " network"}</p>
					</div>
				</div>
			</div>
		</> 
	);
}