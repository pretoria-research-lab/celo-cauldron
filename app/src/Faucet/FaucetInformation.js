import React from 'react';
import QRCode from 'qrcode.react';

export default function FaucetInformation(props) {

    return (
        <div className="row">
          <div className="col-lg-4">
            <h2>{props.network + " Faucet Address"}</h2>
            <hr />
            <p className="address">{props.config ? <a href={props.config.blockExplorer + "/address/" + props.config.faucetAddress}>{props.config.faucetAddress}</a> : ""}</p>
            <hr />
            <div className='qrCode'>                      
                {props.config.faucetAddress ? <QRCode size={150} bgColor="#FFFFFF" fgColor="#2E3338" includeMargin={false} value={props.config.faucetAddress} /> : "" }
            </div>
            <hr />
            <p>{"Balance " + props.faucetBalance + " cGLD"}</p>
            <p>{"Block number " + props.blockNumber}</p>
          </div>
          <div id="how-to-prepare" className="col-lg-8">
              <h2>How To Prepare</h2>
              <hr />
              <ul>                
                <li>Use the table below to request a small slice (0.01) cGLD for testing purposes</li>
                <li>After requesting, the cGLD will cool down for 12 blocks before being served</li>
                <li>Press the confirm button to claim your faucet amount after cooldown</li>
                <li>A maximum of 10 requests can be queued at once</li>                
                <li>If someone leaves without claiming, you may confirm for them at any time</li>
                <li>Only the last 20 requests are visible on this dashboard</li>
              </ul>
              <hr />
              <h5 className="nb">This is a community kitchen - please return any cGLD if you are finished cooking</h5>
          </div>
        </div>
    );
}