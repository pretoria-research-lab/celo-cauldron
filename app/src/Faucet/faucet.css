.technical{
    font-family: "Jost"
}

#loadingLogo {
    width: 150px;
}

thead {
    background-color: var(--celo-canvas-faint-gold);
}

th, td {    
    font-family: "Jost";
    padding: 5px;
}

.address {
    font-family: 'Courier New', Courier, monospace;
    font-size: smaller;
}
 
button {
    width: 150px;
}

p {
    font-size: larger;
}

li {
    font-size:larger;
}

#how-to-prepare {
    text-align: left;
}

#how-to-prepare li {
    font-family: Jost;
}

.nb {
    color: var(--celo-accent-blue);
    font-family: Jost;    
}

.toastInfo {
    font-family: Jost;
    color: var(--celo-dark);
    background-color: var(--celo-green);
    border-bottom: solid 1px var(--celo-dark);
}

.toastWarn {
    font-family: Jost;
    color: var(--celo-dark);
    background-color: var(--celo-gold);
    border-bottom: solid 1px var(--celo-dark);
}

.toastError {
    font-family: Jost;
    color: var(--celo-dark);
    background-color: var(--celo-accent-red);
    border-bottom: solid 1px var(--celo-dark);
}

.toastSuccess {
    font-family: Jost;
    color: var(--celo-dark);
    background-color: var(--celo-green);
    border-bottom: solid 1px var(--celo-dark);
}