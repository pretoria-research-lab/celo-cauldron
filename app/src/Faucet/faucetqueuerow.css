#checkmark {
    max-width: 50px;
}