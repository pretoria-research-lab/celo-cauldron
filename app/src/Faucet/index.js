import React, {Component} from 'react';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import './faucet.css';
import {API_CONFIG} from '../Utils/config';
import {getCurrentBlockNumber} from '../Utils/utils';
import FaucetService from '../FaucetService';
import FaucetQueueTable from './FaucetQueueTable';
import FaucetHeader from './FaucetHeader';
import FaucetInformation from './FaucetInformation';

const faucetService = new FaucetService();
const refreshBlockNumberMillis = 5000;

class Faucet extends Component 
{
  constructor(props){
	  super(props);
      this.state = {faucetRequests: [], loading: false, faucetBalance: 0, config: {}, blockNumber: -1, processing: false};
  };

  notify = (type, message) => {
    if(type==="SUCCESS")
      toast.success(message, {position: toast.POSITION.TOP_RIGHT, className: "toastSuccess"});
    else if(type==="INFO")
      toast.info(message, {position: toast.POSITION.TOP_RIGHT, className: "toastInfo"});
    else if(type==="WARN")
      toast.warn(message, {position: toast.POSITION.TOP_RIGHT, className: "toastWarn"});
    else if(type==="ERROR")
      toast.error(message, {position: toast.POSITION.TOP_RIGHT, className: "toastError"});
    else
      toast.info(message, {position: toast.POSITION.TOP_RIGHT, className: "toastInfo"});
  }

  getBlockNumber = async () => {
    try{
      const nodeProvider = this.state.config.remoteNode;
      const blockNumber = await getCurrentBlockNumber(nodeProvider);
      this.setState({blockNumber: blockNumber});
    }catch(error){
      this.processError(error);
      this.setState({blockNumber: -1});
    }
  }

  getConfig = (network) => {
    const config = API_CONFIG.filter((item) => item.network.toLowerCase() === network.toLowerCase())[0];
    console.log(`Using configuration - ${JSON.stringify(config)}`);
    return config;
  }

  componentDidMount(){
      this.setState({config: this.getConfig(this.props.network)}, () => {
        setInterval(this.getBlockNumber, refreshBlockNumberMillis);
        this.getAllRequests();
      });        
  }

  processError = (error) => {
    if (error.response) {
      console.log(error.response.data);
      console.log(error.response.status);
      console.log(error.response.headers);
      error.response.data.message ? this.notify("ERROR", error.response.data.message) : this.notify("ERROR", error.response.data);
    } else if (error.request) {
      console.log(error.request);
      toast.error(error.request);
    } else {
      console.log(error.message);
      toast.error(error.message);
    }
  }

  claimRequest = (faucetRequest) => {
    this.setState({loading:true}, async () => 
    {
      try{
        const response = await faucetService.claimRequest(this.state.config, faucetRequest.address);
        const claimResult = response.data.claimResult;
        console.log(claimResult);
        this.notify("INFO", "Faucet request completed");
        this.getAllRequests();
        this.setState({loading:false});        
      }
      catch (error) {
        this.processError(error);
        this.setState({loading:false});
      }
    });
  }

  createRequest = (address) => {
    this.setState({loading:true}, async () => 
    {
      try{
        const response = await faucetService.createRequest(this.state.config, address);
        const createResult = response.data;
        console.log(createResult);
        this.getAllRequests();
        this.setState({loading:false});
        this.notify("INFO", "Requested");
      }
      catch (error) {
        this.processError(error);
        this.setState({loading:false});
      }
    });
  }

  getAllRequests = () => {   
    this.setState({loading:true}, async () => {
      try{
        console.log(this.state);
        const response = await faucetService.getAllRequests(this.state.config);
        const faucetRequests = response.data.faucetRequests;
        const faucetBalance = response.data.faucetBalance;
        faucetRequests.forEach(element => {
          element["addressLink"] = this.state.config.blockExplorer + "address/" + element.address;
          element["txLink"] = this.state.config.blockExplorer + "tx/" + element.txId;
        });
        this.setState({faucetRequests}, () => this.setState({faucetBalance}, () => this.setState({loading:false})));
        this.notify("INFO", "Refreshed faucet requests");
      }
      catch (error) {
        toast.warn("Error retrieving faucet requests, please refresh");
        const faucetRequests = [];
        this.setState({faucetRequests}, () => this.setState({loading:false}));
      }
    });
  } 

  render = () => {
    return ( 
      <div className="container-fluid">        
        <FaucetHeader config={this.state.config} network={this.props.network}/>
        <FaucetInformation config={this.state.config} blockNumber={this.state.blockNumber} network={this.props.network} faucetBalance={this.state.faucetBalance}/>
        <FaucetQueueTable config={this.state.config} loading={this.state.loading} blockNumber={this.state.blockNumber} faucetRequests={this.state.faucetRequests} claimRequest={this.claimRequest} createRequest={this.createRequest}/>
        <ToastContainer autoClose={4000}/>
      </div>
    );
  }  
}

export default Faucet;