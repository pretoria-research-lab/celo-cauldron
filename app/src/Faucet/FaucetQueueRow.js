import React from "react";
import PropTypes from "prop-types";
import checkMark from "../assets/10-Check (light-bg).png";
import "./faucetqueuerow.css";

FaucetQueueRow.propTypes = () => { 
	return { 
		faucetRequest: PropTypes.any,
		blockNumber: PropTypes.any,
		config: PropTypes.any,
		onClick: PropTypes.func
	}; 
};	

export default function FaucetQueueRow(props) {

	const { faucetRequest, blockNumber, config, onClick } = props;

	faucetRequest["txIdShort"] = faucetRequest.txId ? faucetRequest.txId.substr(0,10) + "..." + faucetRequest.txId.substr(faucetRequest.txId.length-10,10) : "";
	const blocksUntilReady = config.blocksCooldown - (blockNumber - faucetRequest.createdBlockNumber) < 0 ? 0 : config.blocksCooldown - (blockNumber - faucetRequest.createdBlockNumber);
	const isComplete = faucetRequest.txId;

	return (
		<tr>
			<td className="address"><a href={faucetRequest.addressLink}>{faucetRequest.address}</a></td>
			<td>{faucetRequest.createdBlockNumber}</td>
			<td>{faucetRequest.status}</td>
			{	(blocksUntilReady > 0) && (!isComplete) ?
				<td colSpan="3">{"Cooling down, ready in " + blocksUntilReady + " blocks"}</td>
				:
				<>
					{ 	(!isComplete) ? 
						<>
							<td colSpan="2">Ready to claim</td>
							<td><button onClick={() => onClick(faucetRequest)}>Confirm</button></td>
						</>
						:
						<>
							<td>{faucetRequest.claimedBlockNumber}</td>
							<td className="address"><a href={faucetRequest.txLink}>{faucetRequest.txIdShort}</a></td>
							<td><img id="checkmark" src={checkMark} alt="Claimed check mark" /></td>
						</>
					}
				</>
			}
		</tr>			
	);
}