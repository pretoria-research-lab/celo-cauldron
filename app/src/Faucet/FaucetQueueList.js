
import React, { Component } from 'react';
import { Table } from 'react-bootstrap';
import FaucetQueueRow from './FaucetQueueRow';

export default class FaucetQueueList extends Component {
	
	render = () => {

		const {faucetRequests} = this.props;

		return (
			<div>
				<Table responsive striped bordered hover size="sm">
					<thead className="dark-highlight-bg">
						<tr>
							<th>address</th>
							<th>createdBlockNumber</th>
							<th>status</th>
							<th>claimedBlockNumber</th>
							<th>txId</th>
							<th></th>
						</tr>
					</thead>
					<tbody>
						{faucetRequests.map((fr, i) =><FaucetQueueRow onClick={this.props.claimRequest} key={i} faucetRequest={{...fr}} />)}
					</tbody>
				</Table>
			</div>
		)
	}
}

