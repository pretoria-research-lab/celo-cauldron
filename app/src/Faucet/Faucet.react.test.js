import React from "react";
import Faucet from "./index";
import renderer from "react-test-renderer";

const props = {
	network: "Alfajores"
};

test("Faucet component matches snapshot", () => {
	const component = renderer.create(
		<Faucet {...props}/>,
	);

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot(); 
});