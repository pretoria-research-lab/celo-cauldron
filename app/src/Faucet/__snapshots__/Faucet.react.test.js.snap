// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Faucet component matches snapshot 1`] = `
<div
  className="container-fluid faucet-content"
>
  <div
    className="row centered"
  >
    <div
      className="col-sm-4"
    >
      <div
        className="row page-header"
      >
        <img
          alt="Developer's icon"
          className="celoIcons"
          src="13-Airdrop Green Coin (light bg).png"
        />
        <h2
          className="mt"
        >
          Celo Alfajores Faucet
        </h2>
      </div>
    </div>
    <div
      className="col-sm-8"
    >
      <div
        className="row"
      >
        <p>
          A community funded faucet to receive nominal amounts of cGLD for testing and development on the Celo Alfajores network
        </p>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-lg-4"
    >
      <h3>
        Faucet Address
      </h3>
      <hr />
      <p
        className="address"
      >
        <a
          href="https://alfajores-blockscout.celo-testnet.org//address/0xC02b8b165CC5a2A731E9C2BD6fdF66eaBcfa8663"
        >
          0xC02b8b165CC5a2A731E9C2BD6fdF66eaBcfa8663
        </a>
      </p>
      <hr />
      <div
        className="qrCode"
      >
        <canvas
          height={150}
          style={
            Object {
              "height": 150,
              "width": 150,
            }
          }
          width={150}
        />
      </div>
      <div
        className="technical"
      >
        <p>
          Balance 0 cGLD
        </p>
        <p>
          Block number -1
        </p>
      </div>
    </div>
    <div
      className="col-lg-8"
      id="how-to-prepare"
    >
      <h3>
        How To Prepare
      </h3>
      <hr />
      <ul>
        <li>
          Use the table below to request a 1 cGLD for testing purposes
        </li>
        <li>
          The cGLD will cool down for 12 blocks before being served
        </li>
        <li>
          Press the confirm button to claim your faucet amount after cooldown
        </li>
        <li>
          A maximum of 20 requests can be queued at once
        </li>
        <li>
          If someone leaves without claiming, you may confirm for them at any time
        </li>
        <li>
          Only the last 20 requests are visible on this dashboard
        </li>
      </ul>
      <h5
        className="nb"
      >
        This is a community kitchen - please return any cGLD if you are finished cooking
      </h5>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-lg-12 text-center"
    >
      <img
        alt="Loading..."
        id="loadingLogo"
        src="loading.svg"
      />
    </div>
  </div>
  <div
    className="Toastify"
  />
  <hr />
</div>
`;
